[ONNX export] Support for image normalization in model graph.

francescorubbo opened the pull request:
## Motivation

In order to export models to truly portable ONNX files, the preprocessing steps need to be included in the graph as much as possible.

## Modification

Add option to include the image normalization preprocessing step in the ONNX graph.
The normalization parameters are passed as `img_norm_cfg` via `img_meta` and applied in the forward method.

Fixes #5958

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/open-mmlab/mmdetection?pullRequest=5982) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/open-mmlab/mmdetection?pullRequest=5982) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/open-mmlab/mmdetection?pullRequest=5982) it.</sub>

ZwwWayne commented on mmdet/models/detectors/base.py:
```diff
@@ -164,6 +164,12 @@ def forward(self, img, img_metas, return_loss=True, **kwargs):
+        img_norm_cfg = kwargs.get('img_norm_cfg', None)
+        if img_norm_cfg:
+            mean = torch.tensor(img_norm_cfg['mean'])[None, ..., None, None]
+            std = torch.tensor(img_norm_cfg['std'])[None, ..., None, None]
```
We suggest keeping this logic in lines between 173-175 to restrict its influence.

RunningLeon commented on tools/deployment/pytorch2onnx.py:
```diff
@@ -280,6 +285,10 @@ def parse_args():
+    parser.add_argument(
+        '--normalize-in-graph',
+        action='store_true',
+        help='Whether to include image normalization in ONNX graph.')
```
Could update the doc as well:
https://github.com/open-mmlab/mmdetection/blob/master/docs/tutorials/pytorch2onnx.md#usage

RunningLeon replied:
If a user wants to deploy a model on TensorRT by PyTorch->ONNX->TensorRT, then, this optional feature would have an influence on `onnx2tensorrt.py` tool.

francescorubbo replied:
I've updated the docs and added a `--skip-normalize` option to the `onnx2tensorrt.py` tool. However, I do not have MMCV built with the tensorRT support, so I am not able to test the change myself.

RunningLeon replied:
You could follow the instructions in this doc [onnx2tensorrt](https://github.com/open-mmlab/mmdetection/blob/master/docs/tutorials/onnx2tensorrt.md) to test the logic.

francescorubbo replied:
I'm developing on MacOS, so I can't install TensorRT. Regardless, I've tested that the option is properly picked up by the argparse and that the logic is correct in not including the `normalize_cfg` entry in `input_config` when the `--skip-normalize` is set to `True`.

RunningLeon commented on docs/tutorials/onnx2tensorrt.md:
```diff
@@ -50,6 +51,7 @@ Description of all arguments:
 - `--show`: Determines whether to show the outputs of the model. If not specified, it will be set to `False`.
 - `--verify`: Determines whether to verify the correctness of models between ONNXRuntime and TensorRT. If not specified, it will be set to `False`.
 - `--verbose`: Determines whether to print logging messages. It's useful for debugging. If not specified, it will be set to `False`.
+- `--skip-normalize`: Determins whether to skip normalization of the input image. If not specified, it will be set to `False`. This is useful when the normalization is already included in the ONNX graph.
```
You could add note that this argument should be set to `True` only if `--normalize-in-graph` in `pytorch2onnx.py` is set to `True`.

francescorubbo commented:
@RunningLeon are there additional changes needed for this? Are the tests actually failing (I can't seem to find the logs), or they are killed?

RunningLeon commented:
> @RunningLeon are there additional changes needed for this? Are the tests actually failing (I can't seem to find the logs), or they are killed?

Seems ok to me.

francescorubbo commented:
@RangiLyu what would be the appropriate place to add tests to improve test coverage? As far as I can tell, the tests in `tests/test_onnx` only test inference of model neck and head, not the end to end model that would be needed in this case.

francescorubbo commented:
@RangiLyu thank you for getting back to me. I have added the test and fixed a bug that the test surfaced.
I suspect that we might still be lacking coverage, because the test only covers `mmdet/models/detectors/base.py`, not `mmdet/core/export/pytorch2onnx.py`. Any thought about that?

ZwwWayne commented:
Hi @francescorubbo ,
Thanks for your kind contribution.
Since ONNX will be fully migrated to MMDeploy in the future and we are planning a more general design for the data pre-processing. This PR may not be merged and is thus closed.

OpenMMLab-Assistant005 commented:
Hi @francescorubbo ！We are grateful for your efforts in helping improve this open-source project during your personal time. 
Welcome to join OpenMMLab Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA
If you have a WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤
